File: leetcode/valid_parentheses.py
def isValid(s: str) -> bool:
    stack = []
    opening_brackets = ['(', '{', '[']
    bracket_idx = {
        '(': 0,
        ')': 0,
        '{': 1,
        '}': 1,
        '[': 2,
        ']': 2
    }

    string_list = list(s)
    if len(string_list) <= 1:
        return False

    for bracket in string_list:
        if bracket in opening_brackets:
            stack.append(bracket)
        else:
            if len(stack) > 0 and bracket_idx[stack[-1]] != bracket_idx[bracket]:
                return False
            else:
                if len(stack) > 0:
                    stack.pop()
    
    if len(stack) > 0:
        return False
    
    return True

print(isValid("()"))
print(isValid("()[]{}"))
print(isValid("(]"))
print(isValid("([])"))
print(isValid("["))
print(isValid("]"))

File: leetcode/ValidParentheses.java
import java.util.HashMap;
import java.util.Stack;

public class ValidParentheses {
    class Solution {
        public boolean isValid(String s) {
            Stack<String> stack = new Stack<>();
            HashMap<String, Integer> map = new HashMap<>();

            map.put("[", 1);
            map.put("]", 1);
            map.put("{", 2);
            map.put("}", 2);
            map.put("(", 3);
            map.put(")", 3);

            // String 문자열 각각을 한글자 단위로 변환한다.
            String[] word = s.split("");
            for (String letter : word){
                if (letter.equals("[") || letter.equals("(") || letter.equals("{")){
                    stack.push(letter);
                } else {
                    if (!stack.isEmpty() && map.get(letter) == map.get(stack.peek())){
                        stack.pop();
                    } else {
                        return false;
                    }
                }
            }

            return stack.isEmpty();
        }
    }
}
